File: algols/100-days-of-code/fib_por_exp/main.cpp
#include <bits/stdc++.h>

using namespace std;

struct Mat {
  int arr[2][2];
  Mat() {
    for (int i = 0; i < 2; i++) {
      for (int j = 0; j < 2; j++) {
        arr[i][j] = 0;
      }
    }
  }

  static Mat base() {
    Mat mat = Mat();
    mat.arr[0][0] = 0;
    mat.arr[0][1] = 1;
    mat.arr[1][0] = 1;
    mat.arr[1][1] = 1;
    return mat;
  }

  static Mat multi(Mat &first, Mat &second) {
    Mat valor = Mat();

    for (int i = 0; i < 2; i++) {
      for (int j = 0; j < 2; j++) {
        for (int k = 0; k < 2; k++) {
          valor.arr[i][j] += first.arr[i][k] * second.arr[k][j];
        }
      }
    }
    return valor;
  }
};

Mat fib_base(uint64_t n) {
  if (n == 1) {
    return Mat::base();
  }
  Mat mat = fib_base(n / 2);
  mat = Mat::multi(mat, mat);
  if (n % 2) {
    Mat base = Mat::base();
    mat = Mat::multi(mat, base);
  }

  return mat;
}

uint64_t fib(uint64_t n) {
  if (n == 0) {
    return 0;
  }
  if (n == 1) {
    return 1;
  }

  return fib_base(n).arr[1][1];
}

int main() {
  for (int i = 0; i < 30; i++) {
    int fibonacci = fib(i);
    cout << fibonacci << endl;
  }
  return 0;
}

File: algols/100-days-of-code/fib_por_exp/main.py
import sys

sys.set_int_max_str_digits(0)


def get_base() -> list[list[int]]:
    return [[0, 1], [1, 1]]


def mult(first: list[list[int]], second: list[list[int]]) -> list[list[int]]:
    arr = [[0, 0], [0, 0]]
    for i in range(0, 2):
        for j in range(0, 2):
            for k in range(0, 2):
                arr[i][j] += first[i][k] * second[k][j]
    return arr


def exp_fib(n: int) -> list[list[int]]:
    if n == 1:
        return get_base()

    value = exp_fib(n // 2)
    value = mult(value, value)
    if n % 2 == 1:
        value = mult(value, get_base())
    return value


def fib(n: int) -> int:
    if n == 0:
        return 0
    if n == 1:
        return 1

    value = exp_fib(n)

    return value[1][1]


def fib_linear(n: int) -> int:
    if n == 0:
        return 0
    if n == 1:
        return 1

    a, b = 0, 1

    for _ in range(1, n):
        a, b = b, a + b

    return b


def main():
    print(fib(10_000_000))
    print(fib_linear(10_000_000))


if __name__ == "__main__":
    main()
